tokio/src/io/blocking.rs

```diff
@@ -64,11 +65,11 @@ where
                         return Poll::Ready(Ok(()));
                     }
 
-                    buf.ensure_capacity_for(dst, DEFAULT_MAX_BUF_SIZE);
                     let mut inner = self.inner.take().unwrap();
 
+                    let max_buf_size = cmp::min(dst.remaining(), DEFAULT_MAX_BUF_SIZE);
                     self.state = State::Busy(sys::run(move || {
-                        let res = buf.read_from(&mut inner);
+                        let res = unsafe { buf.read_from(&mut inner, max_buf_size) };
```

Comment:
We need to also mark `Blocking::new` as unsafe and require that `self.inner` satisfies the requirement. Otherwise this unsafe block isn't really correct.